cpp/include/raft/spatial/knn/detail/ball_cover.cuh

```diff
@@ -539,6 +617,91 @@ void rbc_knn_query(raft::resources const& handle,
                     perform_post_filtering);
 }
 
+template <typename value_idx,
+          typename value_t,
+          typename value_int  = std::int64_t,
+          typename matrix_idx = std::int64_t>
+void compute_landmark_dists(raft::resources const& handle,
+                            const BallCoverIndex<value_idx, value_t, value_int, matrix_idx>& index,
+                            const value_t* query_pts,
+                            value_int n_query_pts,
+                            value_t* R_dists)
+{
+  // compute distances for all queries against all landmarks
+  // index.get_R() -- landmark points in row order (index.n_landmarks x index.k)
+  // query_pts     -- query points in row order (n_query_pts x index.k)
+  // TODO(mfoerste4): sice we cast to int, we shall assert whether matrices can be idexed with int
+  raft::distance::pairwise_distance<value_t, int>(handle,
+                                                  query_pts,
+                                                  index.get_R().data_handle(),
+                                                  R_dists,
+                                                  n_query_pts,
+                                                  index.n_landmarks,
+                                                  index.n,
+                                                  index.get_metric());
+}
+
+/**
+ * Performs a knn query against an index. This assumes the index has
+ * already been built.
+ * Modified version that takes an eps as threshold and outputs to a dense adj matrix (row-major)
+ * we are assuming that there are sufficiently many landmarks
+ */
+template <typename value_idx = std::int64_t,
+          typename value_t,
+          typename value_int  = std::int64_t,
+          typename matrix_idx = std::int64_t,
+          typename distance_func>
+void rbc_eps_nn_query(raft::resources const& handle,
+                      const BallCoverIndex<value_idx, value_t, value_int, matrix_idx>& index,
+                      const value_t eps,
+                      const value_t* query,
+                      value_int n_query_pts,
+                      bool* adj,
+                      value_idx* vd,
+                      distance_func dfunc)
+{
+  ASSERT(index.is_index_trained(), "index must be previously trained");
+
+  rmm::device_uvector<value_t> R_dists(index.n_landmarks * n_query_pts,
```

Comment:
We can use adevice matrix
```suggestion
auto R_dists = raft::make_device_bahcix<value_t, matrix_idx>(res, n_landbarks, n_query_pts);
```
(just remember to update data pointers below to `R_dists.data_handle()`.